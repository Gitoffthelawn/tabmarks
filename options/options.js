const options = {

  rootFolderName: null,

  init() {
    document.addEventListener('click', this.handleEvent.bind(this));
    document.addEventListener('submit', this.handleEvent.bind(this));
    this.port = browser.runtime.connect({ name: 'options' });
    this.port.onMessage.addListener(this.handleMessage.bind(this));
  },

  handleEvent(e) {
    if (e.target.id === 'nameEditButton') {
      this.editName(true);
      this.nameInput.focus();
    } else if (e.target.id === 'nameCancelButton') {
      this.editName(false);
      this.nameInput.value = this.rootFolderName;
    } else if (e.target.id === 'nameForm') {
      e.preventDefault();
      this.saveRootFolderName();
      this.editName(false);
    }
  },

  handleMessage(message) {
    switch (message.message) {
      case 'updateRootFolderName':
        this.updateRootFolderName(message.rootFolderName);
        break;
      default:
        console.error('Received unknown message:', message);
    }
  },

  updateRootFolderName(rootFolderName) {
    this.rootFolderName = rootFolderName;
    this.nameInput.value = rootFolderName;
    this.editName(false);
  },

  saveRootFolderName() {
    const name = this.nameInput.value;
    if (name.trim() === this.rootFolderName) {
      this.nameInput.value = this.rootFolderName;
      return;
    }
    tm.bookmarks.renameRootFolder(name);
  },

  get nameInput() {
    return document.querySelector('#rootFolderName');
  },

  get nameEditButton() {
    return document.querySelector('#nameEditButton');
  },

  get nameCancelButton() {
    return document.querySelector('#nameCancelButton');
  },

  get nameSaveButton() {
    return document.querySelector('#nameSaveButton');
  },

  editName(enabled) {
    this.nameInput.disabled = !enabled;
    this.nameEditButton.style.display = enabled ? 'none' : '';
    this.nameCancelButton.style.display = enabled ? '' : 'none';
    this.nameSaveButton.style.display = enabled ? '' : 'none';
  },

};

options.init();
