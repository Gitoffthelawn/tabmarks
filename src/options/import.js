if (!window.tm) window.tm = {};

tm.import = {

  importTabGroupsJson(json) {
    let data;
    try {
      data = JSON.parse(json);
      if (Array.isArray(data.windows)) {
        return data.windows.reduce((result, windowData) =>
          result.then(() => this.importWindows(windowData)),
        Promise.resolve());
      }
      return Promise.reject('Invalid JSON structure');
    } catch (e) {
      return Promise.reject('Unable to parse JSON', e);
    }
  },

  importWindows(data) {
    const groups = this.getGroupsData(data);
    const tabs = this.getTabsData(data);

    if (groups) {
      groups.forEach((group, i) => {
        groups[i].tabs = tabs.filter(t => t.groupId === group.id);
      });
      return this.importGroups(groups);
    }

    return Promise.resolve();
  },

  importGroups(groups) {
    return groups.reduce((result, group) =>
      result.then(() => tm.bookmarks.createFolder(group.title)
        .then(folder => this.importTabs(group.tabs, folder)),
      error => console.error(`Could not create folder for group "${group.title}"`, error)),
    Promise.resolve());
  },

  importTabs(tabs, folder) {
    return tabs.reduce((result, tab) =>
      result.then(() => tm.bookmarks.create(tab.title, tab.url, folder.id))
        .catch((error) => {
          console.error(`Could not create bookmark "${tab.url}" in group "${folder.id}"`, error);
        }),
    Promise.resolve());
  },

  getGroupsData(data) {
    if (data && data.extData && data.extData['tabview-group']) {
      try {
        const groups = JSON.parse(data.extData['tabview-group']);
        return Object.keys(groups).map(id => ({ id: parseInt(id, 10), title: groups[id].title }));
      } catch (e) {
        return null;
      }
    }
    return null;
  },

  getTabsData(data) {
    if (Array.isArray(data.tabs)) {
      return data.tabs
        .map(this.getTabData)
        .filter(t => t && t.url && t.url.indexOf('about:') !== 0);
    }
    return [];
  },

  getTabData(data) {
    if (Array.isArray(data.entries) && data.entries.length > 0) {
      const entry = data.entries[0];
      if (data && data.extData && data.extData['tabview-tab']) {
        try {
          const groupData = JSON.parse(data.extData['tabview-tab']);
          if (groupData && groupData.groupID != null) {
            return {
              title: entry.title,
              url: entry.url,
              groupId: parseInt(groupData.groupID, 10),
            };
          }
        } catch (e) {
          return null;
        }
      }
    }
    return null;
  },

};
