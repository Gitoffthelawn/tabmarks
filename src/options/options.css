.options-section {
  padding: 0;
}

.options-item {
  display: flex;
  align-items: start;
  margin: 0;
  border-top: 1px solid #c1c1c1;
}

.options-label {
  flex: 1;
  padding: 6px;
  text-align: left;
}

.options-content {
  flex: 2;
  display: flex;
  padding: 6px;
}

.options-content-field {
  flex: auto;
}

.options-content-field > * {
  width: 100%;
}

.options-content-actions {
  flex: none;
  padding-left: 6px;
}

.spinner {
  width: 1em;
  height: 1em;
  margin-right: 3px;
  background-image: url(../../assets/icons/tabmarks.svg);
  background-size: contain;
  animation: spin 3000ms linear infinite;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
