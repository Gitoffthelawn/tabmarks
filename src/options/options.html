<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8">
    <link rel="stylesheet" href="options.css">
  </head>

  <body class="browser-style">
    <div class="options-section panel-section panel-section-formElements">

      <form id="nameForm" class="options-item panel-formElements-item">
        <label class="options-label browser-style-label">Root folder name</label>
        <div class="options-content">
          <div class="options-content-field browser-style">
            <input id="rootFolderName"
                   type="text">
          </div>
          <div class="options-content-actions">
            <button id="nameEditButton"
                    type="button" class="browser-style">Edit</button>
            <button id="nameCancelButton"
                    type="button" class="browser-style">Cancel</button>
            <button id="nameSaveButton"
                    type="submit" class="browser-style">Save</button>
          </div>
        </div>
      </form>

    </div>

    <script src="../background/helpers/bookmarks.js"></script>
    <script src="options.js"></script>
  </body>

</html>
