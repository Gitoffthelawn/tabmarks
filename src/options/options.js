import bookmarksHelper from '../helpers/bookmarks';
import importHelper from '../helpers/import';

export default {

  rootFolderName: null,

  init() {
    document.addEventListener('click', this.onClick.bind(this));
    document.addEventListener('submit', this.onSubmit.bind(this));

    this.port = browser.runtime.connect({ name: 'options' });
    this.port.onMessage.addListener(this.onMessage.bind(this));
  },

  onClick(e) {
    if (e.target.id === 'nameEditButton') {
      this.editName(true);
      this.nameInput.focus();
    } else if (e.target.id === 'nameCancelButton') {
      this.editName(false);
      this.nameInput.value = this.rootFolderName;
    } else if (e.target.id === 'importMessageButton') {
      this.showImportContent();
    }
  },

  onSubmit(e) {
    e.preventDefault();

    if (e.target.id === 'nameForm') {
      this.saveRootFolderName();
      this.editName(false);
    } else if (e.target.id === 'importForm') {
      this.import();
    }
  },

  onMessage(message) {
    switch (message.message) {
      case 'updateRootFolderName':
        this.updateRootFolderName(message.rootFolderName);
        break;
      default:
        console.error('Received unknown message:', message);
    }
  },

  updateRootFolderName(rootFolderName) {
    this.rootFolderName = rootFolderName;
    this.nameInput.value = rootFolderName;
    this.editName(false);
  },

  saveRootFolderName() {
    const name = this.nameInput.value;
    if (name.trim() === this.rootFolderName) {
      this.nameInput.value = this.rootFolderName;
      return;
    }
    bookmarksHelper.renameRootFolder(name);
  },

  editName(enabled) {
    this.nameInput.disabled = !enabled;
    this.nameEditButton.style.display = enabled ? 'none' : '';
    this.nameCancelButton.style.display = enabled ? '' : 'none';
    this.nameSaveButton.style.display = enabled ? '' : 'none';
  },

  import() {
    this.showImportProgress();
    importHelper.importTabGroupsJson(this.importField.value)
      .then(() => {
        this.importField.value = '';
        this.showImportMessage('Import finished.');
        this.port.postMessage({ message: 'refreshGroups' });
      }, (error) => {
        this.showImportMessage(`An error occurred: ${error}`);
      });
  },

  showImportContent() {
    this.importContent.style.display = '';
    this.importProgress.style.display = 'none';
    this.importMessage.style.display = 'none';
  },

  showImportProgress() {
    this.importContent.style.display = 'none';
    this.importProgress.style.display = '';
    this.importMessage.style.display = 'none';
  },

  showImportMessage(message) {
    this.importContent.style.display = 'none';
    this.importProgress.style.display = 'none';
    this.importMessage.querySelector('.text').textContent = message;
    this.importMessage.style.display = '';
  },

  get nameInput() {
    return document.getElementById('rootFolderName');
  },

  get nameEditButton() {
    return document.getElementById('nameEditButton');
  },

  get nameCancelButton() {
    return document.getElementById('nameCancelButton');
  },

  get nameSaveButton() {
    return document.getElementById('nameSaveButton');
  },

  get importButton() {
    return document.getElementById('importButton');
  },

  get importField() {
    return document.getElementById('importField');
  },

  get importContent() {
    return document.getElementById('importContent');
  },

  get importProgress() {
    return document.getElementById('importProgress');
  },

  get importMessage() {
    return document.getElementById('importMessage');
  },

};
