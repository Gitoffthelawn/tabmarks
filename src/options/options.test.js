import options from './options';
import importHelper from '../helpers/import';

jest.mock('../helpers/import', () => ({
  importTabGroupsJson: jest.fn().mockImplementation(() => Promise.resolve()),
}));

describe('options', () => {
  describe('#import', () => {
    beforeEach(() => {
      options.showImportProgress = jest.fn();
      options.showImportMessage = jest.fn();
      delete options.importField;
      options.importField = { value: '{ "foo": "bar" }' };
      options.port = { postMessage: jest.fn() };
    });

    test('imports JSON data and displays success message', () =>
      options.import().then(() => {
        expect(importHelper.importTabGroupsJson.mock.calls).toMatchSnapshot();
        expect(options.showImportProgress).toBeCalled();
        expect(options.showImportMessage.mock.calls).toMatchSnapshot();
      }));

    test('imports JSON data and displays failed folder/bookmark creations', () => {
      importHelper.importTabGroupsJson = jest.fn().mockImplementation((json, errorCallback) => {
        errorCallback({ type: 'folder', name: 'Group A', error: 'Folder creation failed' });
        errorCallback({ type: 'bookmark', name: 'https://developer.mozilla.org/en-US/', error: 'Bookmark creation failed' });
        errorCallback({ type: 'bookmark', name: 'https://github.com/', error: 'Bookmark creation failed' });
        return Promise.resolve();
      });
      return options.import().then(() => {
        expect(importHelper.importTabGroupsJson.mock.calls).toMatchSnapshot();
        expect(options.showImportProgress).toBeCalled();
        expect(options.showImportMessage.mock.calls).toMatchSnapshot();
      });
    });

    test('displays error on complete import failure', () => {
      importHelper.importTabGroupsJson = jest.fn().mockImplementation(() => Promise.reject('Bad'));
      return options.import().then(() => {
        expect(importHelper.importTabGroupsJson.mock.calls).toMatchSnapshot();
        expect(options.showImportProgress).toBeCalled();
        expect(options.showImportMessage.mock.calls).toMatchSnapshot();
      });
    });
  });
});
