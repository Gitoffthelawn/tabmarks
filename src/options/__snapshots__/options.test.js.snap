// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options #import displays error on complete import failure 1`] = `
Array [
  Array [
    "{ \\"foo\\": \\"bar\\" }",
    [Function],
  ],
]
`;

exports[`options #import displays error on complete import failure 2`] = `
Array [
  Array [
    "An error occurred: Bad",
  ],
]
`;

exports[`options #import imports JSON data and displays failed folder/bookmark creations 1`] = `
Array [
  Array [
    "{ \\"foo\\": \\"bar\\" }",
    [Function],
  ],
]
`;

exports[`options #import imports JSON data and displays failed folder/bookmark creations 2`] = `
Array [
  Array [
    "Import finished. (failed to create: folder Group A, bookmark https://developer.mozilla.org/en-US/, bookmark https://github.com/)",
  ],
]
`;

exports[`options #import imports JSON data and displays success message 1`] = `
Array [
  Array [
    "{ \\"foo\\": \\"bar\\" }",
    [Function],
  ],
]
`;

exports[`options #import imports JSON data and displays success message 2`] = `
Array [
  Array [
    "Import finished.",
  ],
]
`;
